### DISCUSSION TITLE
[Bug] Explicit layout with record struct

### DISCUSSION THREAD

**Discussion by Xyncgas** [fsharp/fslang-suggestions#1233] [General]
For example
```F#
[<Struct; StructLayout(LayoutKind.Explicit)>]
 type Dimension2D = 
            {
                [<FieldOffsetAttribute(0)>]
                Memory : Vector2
                [<FieldOffsetAttribute(0)>]
                Width : float32
                [<FieldOffsetAttribute(4)>]
                Height : float32
            }
            with
            static member Zero = { Memory = Vector2.Zero; Width = 0f; Height = 0f; }
```
The point of explicit layout is you can change the memory by changing width / height, or the opposite way by using System.Numerics.Vector2 methods to change the memory

However I noticed that, while it's already hard to create a record with only memory is set
```F#
{Dimension2D.Memory = Vector2.Zero} //Alone doesn't work
```
If I were to do :
```F#
{ Dimension2D.Zero with Memory = Vector2.Multiply(aVector, value) } //Bug : the value remains zero after this no matter what you do
```
Where ```AnotherMemory``` is not zero and so is value, the result is zero for some reason

But if I were to do
```F#
let v = Vector2.Multiply(aVector, value)
{ Dimension2D.Zero with Memory = v; Width = aVector.X; Height = aVector.Y } //Works
```

It gives me a feeling that either I can't have record struct with explicit layout in F# like I can in C#, or there is a bug with it

I wanted to do this because if I store my value in Vector struct they get hardware accelerated sometimes. Performance is important to F# aye, so I am hoping I can write these kinds of codes

**Comment by vzarytovskii** [comment]
I would imagine this is by design - a default behaviour of explicitly copying fields in `with` syntax

**Update:** As you can see [here](https://sharplab.io/#v2:EYLgtghglgdgNAGxAMwM4B8D2AHApjAAgGUBPVAF1zADoAlAVxnKjF2oEkncAnHIngG5QAxrlQBYAFA58xMpRoA5eq24iJkgNoAeIuW71h5ANzF9h8gBkIJTPXIAKa7fsBpWABNqAUQAe2BBEocgBKAD4AXSlyEjwCABEWfFQoTBgAJniCAF4pAnyCAG88gtKdADEoXAQPAHlkNFxHAAZwqMlS0oBZKkxuEgIQAgA1XCM+9JLOggqqmvrGlrapzoB1KA9yAAtBgmQETAhyAGZJjumZ7UrquobUJocAFmXz6YAJXCgAcy3yXf3DiczqUAL4rADuwS2KwoRxEBFYYGAPAIAC0eJgckUCD0wH0BtkRmNyBNqOjeKZ1psdoTmshTB9vr8sXTTGDJAgmgQYLhwaNhFjCgkkjAUmlMmSMQRIdscb1+lj+STuOlqF16AhmAESA4lRMHKrkCE4ETxiqHABWahGkIEdnYNRMZCEAAkACJCjy+WMQW6gA=), the creation is explicit:
```csharp
copyOfStruct@15 = new @_.Dimension2D(Vector2.Zero, 0f, 0f);
copyOfStruct@15-1 = @_.copyOfStruct@15;
newVec@15 = new @_.Dimension2D(Vector2.Multiply(new Vector2(2f), new Vector2(5f)), copyOfStruct@15-1.Width@, copyOfStruct@15-1.Height@);
```

**Reply by vzarytovskii** [reply]
Note, that C#'s behaviour is different, it does only set one property/field: https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEAMACLARgDoAldBKMAU2IEkYFrkB7ABwGVmA3KAY2oBnANwBYAFC4CJAHKoayfiIkSixIgE4AFABEq1GIKgsYAJh3EAYixZaAlHbHiJAbQ4I0fBABkIATxZUBC1fAKCAaVgAE2IAUQAPNgAbfigEOwBdCWRqPhZkKLxBD1QvPD0aQ2MzHQkAbwk8JrwXSyhqJKiAeQAzHsFqYJxMxuasAGY8ADVchHzTPABZajB8vydmvFGm1vbO3v7BrWGs8U2JvB6klggEPAB1KCiEAAsN5u2Wto7uvoHggAsIzOY0mVxudwAEtQoABzF4Id5NT4XIgANnK+iqJnMeAAWswWHg6nhYYNhHgAL54AC8n02MGoAHdMZUjDidMSlis1rTprN5sQCaw4A8nq8+ThRdC4QjJVSkVsQcjlQRJujWQZ2TU8NZbHZaQA+TXYmpCwl4JlpF5c5arZB+PkzLyCxaoJKUZJ+LSMlnOubIUxaUx2UW+/kuwNaACsDgVKnElITQA==

**Reply by Xyncgas** [reply]
Ahh yes, the values I set turns into 0 because I made my struct layout explicit and both ```Width``` and ```Height``` are within the range of ```Memory```'s memory. Therefore by setting ```Width``` and ```Height``` to 0 my ```Memory``` becomes 0.

```record set - > other fields copied -> my input gets overwritten / but if other fields copied -> now set record with my input -> I overwrite other fields```

So it's a simple fix, copy the value first, then do whatever the user wants to do with the ```with``` keyword next. Just swap the order a little bit. This way, the values users are providing (are made sure) overwrites whatever record you were copying. The behavior stays the same if you are not using explicit layout, but if you are then it makes your value to overwrite the old value and the other values that's within the range of the layout of the value you are setting which fixes the problem for me without creating problem for the other people.

**Reply by Xyncgas** [reply]
Also, it's a bug not a feature because we are seeing conflicts in how record works and how this is behaving, the former was intended while this is..

**Reply by Xyncgas** [reply]
If I modify something in my explicit laid out record data structure I would like to keep everything unchanged first, then change what I want to change, then keep that change
